source/vir/src/split_expression.rs

```diff
@@ -0,0 +1,546 @@
+        }
+        ExpX::Bind(bnd, e1) => {
+            // In the arg_map, remove names that will be binded locally
+            let mut arg_map = arg_map.clone();
+            arg_map.retain(|name, _| {
+                let binded_names: Vec<Ident> = match &bnd.x {
+                    BndX::Let(binders) => binders.iter().map(|b| b.name.clone()).collect(),
+                    BndX::Quant(_, binders, _) => binders.iter().map(|b| b.name.clone()).collect(),
+                    BndX::Choose(binders, _, _) => binders.iter().map(|b| b.name.clone()).collect(),
+                    BndX::Lambda(binders) => binders.iter().map(|b| b.name.clone()).collect(),
+                };
+                !binded_names.iter().any(|bname| **bname == **name)
+            });
+            let arg_map = arg_map;
+            let e1_replaced = subsitute_argument(e1, &arg_map)?;
+            let bnd_replaced: Bnd = match &bnd.x {
+                BndX::Let(binders) => {
+                    let mut new_binders: Vec<Binder<Exp>> = vec![];
+                    for old_b in &**binders {
+                        let new_b = BinderX {
+                            name: old_b.name.clone(),
+                            a: subsitute_argument(&old_b.a, &arg_map)?,
+                        };
+                        new_binders.push(Arc::new(new_b));
+                    }
+                    Spanned::new(bnd.span.clone(), BndX::Let(Arc::new(new_binders)))
+                }
+                BndX::Quant(quant, bndrs, _trigs) => Spanned::new(
+                    bnd.span.clone(),
+                    BndX::Quant(quant.clone(), bndrs.clone(), Arc::new(vec![])),
+                ),
+                _ => {
+                    return Err((
+                        exp.span.clone(),
+                        format!("Unsupported binder during subsitution"),
+                    ));
+                }
+            };
+            let new_exp = ExpX::Bind(bnd_replaced, e1_replaced);
+            SpannedTyped::new(&exp.span, &exp.typ, new_exp)
+        }
+        ExpX::Const(_) => exp.clone(),
+        _ => {
+            return Err((exp.span.clone(), format!("Unsupported expression during subsitution")));
+        }
+    };
+    Ok(result)
+}
+
+fn is_bool_type(t: &Typ) -> bool {
+    match &**t {
+        TypX::Bool => true,
+        TypX::Boxed(b) => is_bool_type(b),
+        _ => false,
+    }
+}
+
+fn is_same_type(t1: &Typ, t2: &Typ) -> bool {
+    match (&**t1, &**t2) {
+        (TypX::Bool, TypX::Bool)
+        | (TypX::Int(_), TypX::Int(_))
+        | (TypX::Tuple(_), TypX::Tuple(_))
+        | (TypX::Lambda(_, _), TypX::Lambda(_, _))
+        | (TypX::Datatype(_, _), TypX::Datatype(_, _))
+        | (TypX::TypParam(_), TypX::TypParam(_))
+        | (TypX::TypeId, TypX::TypeId)
+        | (TypX::Air(_), TypX::Air(_)) => true,
+        (TypX::Boxed(b1), TypX::Boxed(b2)) => is_same_type(b1, b2),
+        _ => false,
+    }
+}
```

Comment:
There's already a similar function ast_util, `types_equal`. This function looks like it computes something different, only comparing the root nodes of the type trees, and it's not clear to me if that's a mistake or if it's intentional. If it _is_ intentional, it isn't clear to me at all why you'd want to compute this, so please document it.

Reply:
Thanks for the review!  You are right, I should use `types_equal`.